File: 46. Permutations/solution.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        sol = []
        # edge case
        if not nums:
            return sol
        
        numset = set(nums)
        
        # backtrack
        return self.backtrack(numset)
        
    # O(n!) for backtracking over numset of size n
    def backtrack(self, numset):    
        sol = []
        
        # base case
        if len(numset) == 1:
            num = numset.pop()
            sol.append([num])
            numset.add(num)
        else:
            # loop over n numbers
            nums = list(numset)
            for num in nums:
                numset.remove(num)
                sub = self.backtrack(numset)
                for s in sub:
                    s.append(num)
                    sol.append(s)
                numset.add(num)
        return sol
            

File: 46. Permutations/solution.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        Set<Integer> visited = new HashSet<Integer>();
        return backtrack(nums, visited);
    }
    
    public List<List<Integer>> backtrack(int[] nums, Set<Integer> visited) {
        List<List<Integer>> sol = new ArrayList<List<Integer>>();
        if (nums.length == visited.size()) {
            sol.add(new ArrayList<Integer>());
            return sol;
        } else {
            for (int num: nums) {
                if (!visited.contains(num)) {
                    visited.add(num);
                    List<List<Integer>> subSol = backtrack(nums, visited);
                    for (List<Integer> s: subSol) {
                        s.add(num);
                        sol.add(s);
                    }
                    visited.remove(num);
                }
            }
        }
        
        return sol;
    }
}
